### ISSUE TITLE
Typing ModelForm causes not subscriptable error

### ISSUE THREAD

**Issue by mschoettle** [typeddjango/django-stubs#1389]
# Bug report

## What's wrong

I changed `disallow_any_generics` to `True` and going through all the errors reported by `mypy`. One of them is on `DeleteView` and related views that expects two type parameters.

I changed `class FooDeleteView(DeleteView)` to `class FooDeleteView(DeleteView[Foo, ModelForm[Foo]])`.

While mypy is now happy, running the code causes the following error:

```python
TypeError: 'ModelFormMetaclass' object is not subscriptable
```

Might be related to #703 

## How is that should be

No error should be caused.

I am also wondering if definining it as `DeleteView[Foo]` would be possible to be sufficient since `ModelForm[Foo]` can be derived from it.

## System information

- OS:
- `python` version: 3.10.10
- `django` version: 4.1.7
- `mypy` version: 1.0.1
- `django-stubs` version: 1.15.0
- `django-stubs-ext` version: 0.7.0

**Comment by mschoettle** [comment]
It also happens when providing type arguments for a form itself:

```python
class FooModelForm(forms.ModelForm[Foo])
```

**Comment by intgr** [comment]
> TypeError: 'ModelFormMetaclass' object is not subscriptable

Yep, I think these classes may need to be added to our monkeypatching, similar to #703 as you point out.

> I am also wondering if definining it as DeleteView[Foo] would be possible to be sufficient since ModelForm[Foo] can be derived from it.

This can be done with a type alias and renaming the original class. Something similar is done with `QuerySet`, if you write `QuerySet[User]` then that gets expanded to `_QuerySet[User, User]`.

Although that is quite a hack IMO, it requires renaming the class in type stubs. In addition to ugliness in diagnostic messages, this makes PyCharm IDE integrations worse: PyCharm does not understand that Django's `QuerySet` class is the same as `_QuerySet` in stubs.

**Comment by mschoettle** [comment]
Thanks for the quick response! I'll see if I can get a PR cooked up for this.

> This can be done with a type alias and renaming the original class. Something similar is done with QuerySet, if you write QuerySet[User] then that gets expanded to _QuerySet[User, User].
> 
> Although that is quite a hack IMO, it requires renaming the class in type stubs. In addition to ugliness in diagnostic messages, this makes PyCharm IDE integrations worse: PyCharm does not understand that Django's QuerySet class is the same as _QuerySet in stubs.

Ah, I've seen that `_QuerySet` reference in VS Code. I don't know if it's shown in PyCharm but when inheriting from a generic class like `QuerySet` VS Code (or mypy?) does not show which type parameters are expected. Is that something that is possible to do?

I agree with you that its better to be explicit about the types. It also is in line with the Zen of Python ("explicit is better than implicit").

**Comment by intgr** [comment]
> I agree with you that its better to be explicit about the types.

If it could be done without hacks, I think we should do it. But for now, I'd prefer the explicit approach too.

I did a Google search and whaddya know, there's already a PEP to implement this properly: [PEP 696 – Type defaults for TypeVarLikes](https://peps.python.org/pep-0696/).

> when inheriting from a generic class like QuerySet VS Code (or mypy?) does not show which type parameters are expected. Is that something that is possible to do?

I think VS Code *could* support it with a type alias like `QuerySet` as well, it's probably an unimplemented feature there.

PyCharm doesn't show typevars in such a way at all.

**Comment by mschoettle** [comment]
My apologies. I was trying to fix it and noticed that I was missing the call to `django_stubs_ext.monkeypatch()`. Turns out it works already without any changes.

I might add a FAQ entry to point to the right places though if you are good with that.
